### DISCUSSION TITLE
Cursor Blinking Style

### DISCUSSION THREAD

**Discussion by austincondiff** [migueldeicaza/SwiftTerm#279] [Ideas]
We use SwiftTerm in [CodeEdit](https://github.com/CodeEditApp/CodeEdit) and we are wanting to offer similar cursor blinking styles [seen in VS Code](https://chrisoncode.hashnode.dev/animating-your-vs-code-cursor-with-cursor-blinking) both in our editor which uses [STTextView](https://github.com/krzyzanowskim/STTextView), and in our Terminal. The blink styles offered by VS Code are: 

| Style | Preview |
| - | - |
| Blink | <img src="https://user-images.githubusercontent.com/806104/223346401-6ec068a6-8c45-4ad5-ad6c-7373521feff2.gif" alt="blink" width="256"> |
| Smooth | <img src="https://user-images.githubusercontent.com/806104/223346401-6ec068a6-8c45-4ad5-ad6c-7373521feff2.gif" alt="blink" width="256"> |
| Phase | <img src="https://user-images.githubusercontent.com/806104/223346733-9eb6b111-d523-41ae-b020-a2bca7bb9162.gif" alt="blink smooth" width="256"> |
| Expand | <img src="https://user-images.githubusercontent.com/806104/223346815-6936cc2d-f66d-492c-a5ae-c5d4b520e6f0.gif" alt="blink expand" width="256"> |
| Solid | <img src="https://user-images.githubusercontent.com/806104/223346881-ec8a34d6-1221-4ecc-a1d7-3c6b4e9f8f00.gif" alt="blink" width="256"> |

We might experiment to offer other variations of these (maybe phase+expand for example).

After some discussion with @bombardier200 and a few other of our contributors we may be able to contribute more as we have a growing interest in this projects development. This seems like it would need to be done at the SwiftTerm level the way I see it currently. We could include this in SwiftTerm, but we'd like to be consistent in our implementation between our editor and terminal so it may be good to make this possible externally. 

I’d imagine there are many possible ways this can be configured. We might provide the following configurations, we could even use these configurations internally to create presets.

| Configuration                 | Default (Blink) | Smooth | Phase | Expand | Solid |
| ----------------------------- | --------------- | ------ | ----- | ------ | ----- |
| blink on duration             | 0.5s            | 0.5s   | 0.5s  | 0.5s   | 0.5s  |
| blink off duration            | 0.5s            | 0.5s   | 0.5s  | 0.5s   | 0.5s  |
| transition in duration        | 0s              | 0.05s  | 0.2s  | 0.2s   | 0s    |
| transition out duration       | 0s              | 0.05s  | 0.2s  | 0.2s   | 0s    |
| opacity at on state           | 1               | 1      | 1     | 1      | 1     |
| opacity at off state          | 0               | 0      | 0     | 0      | 1     |
| vertical scale at on state    | 1               | 1      | 1     | 1      | 1     |
| vertical scale at off state   | 1               | 1      | 1     | 0      | 1     |
| horizontal scale at on state  | 1               | 1      | 1     | 1      | 1     |
| horizontal scale at off state | 1               | 1      | 1     | 1      | 1     |

@migueldeicaza What are your thoughts? Is this something we could build into SwiftTerm?

**Comment by migueldeicaza** [comment]
Hello,

In general, I think that this is a good idea, there are a few things to keep in mind.

There is already a set of cursor styles in SwiftTerm which correspond to the settings that users can configure via escape sequences.  Look at the `func cmdSetCursorStyle (_ pars: [Int], _ collect: cstring)` method to see where this is done.  These conflict, but we could think of Blink, Smooth, Phase and Expand as rendering styles for blink.   So these could be implemented as an AppKit/UIKit "blinkStyle" specific feature that is used when the user selects the blink style.

The secondary issue that you will run into is that the cursor rendering in SwiftTerm is not great (https://github.com/migueldeicaza/SwiftTerm/issues/219).   It is just a UIView on top of the terminal.   This makes it very easy to implement the blinking, but also it means that you are limited in what the cursor can do.  And in fact, you will notice that a fully filled cursor tends to completely obscure the text, rather than providing the nice behavior that your screenshots show.

There are two possible solutions to this problem:

1. You could make it so that the CaretView is not just an empty box, but one that fetches the current character and draws the current character as expected.   This is slightly complicated due to type setting issues (you would need to either keep the CTRun cached, or you would need to compute the CTRun for it).   This has the advantage that if properly done, it would work on iOS and macOS.   This might have an ugly performance hit due to having to compute the layout for every cursor movement.

2. At least on MacOS it is possible to apply CoreImage filters to NSViews, and it might be possible to use one of those CIFilters to achieve an effect that does not entirely obscure the underlying text.   I did not implement this purely because my main use is on iOS, and it did not work there.

Option (1) has I suspect assorted corner cases, but also would allow SwiftTerm to get support for cursor foreground/background colors, which I currently ignore.  Option (2) is likely faster for you to get up to speed, and it is an acceptable band-aid for the short term, and can always be revisited.

**Reply by migueldeicaza** [reply]
I fixed the second cursor issue - I can finally sleep at night

**Reply by migueldeicaza** [reply]
After I implemented the new cursor support all of these are possible.   All that we need now is to have a way of passing the desired animations on the API and have some reasonable mapping to the terminal status codes (map commands that set it to one of these things and respond with something reasonable when queried)

**Reply by bombardier200** [reply]
That's great! I can't wait to try it out!

**Reply by austincondiff** [reply]
Fantastic! Nice work @migueldeicaza. How did you get the current character rendering? Are we incurring a big performance hit?

**Reply by migueldeicaza** [reply]
The best of all worlds: I now render the exact character when the cursor moves, so this means the buffer never has to be re-rendered.  So this is super cheap to do!

I was afraid it was going to be a lot of work, but it was so simple.
